File: implementations/ruby/sorting/cocktail/cocktail_sort.rb
def cocktail_shaker_sort(values)
  a = values.dup
  start = 0
  finish = a.length - 1
  swapped = true
  while swapped
    swapped = false
    (start...finish).each do |i|
      if a[i] > a[i + 1]
        a[i], a[i + 1] = a[i + 1], a[i]
        swapped = true
      end
    end
    break unless swapped
    swapped = false
    finish -= 1
    finish.downto(start + 1) do |i|
      if a[i - 1] > a[i]
        a[i - 1], a[i] = a[i], a[i - 1]
        swapped = true
      end
    end
    start += 1
  end
  a
end
raise 'primary' unless cocktail_shaker_sort([5, 1, 4, 2, 8, 0, -3]) == [-3, 0, 1, 2, 4, 5, 8]
raise 'duplicate' unless cocktail_shaker_sort([3, 3, 2, 1]) == [1, 2, 3, 3]
puts 'ruby cocktail ok'

File: implementations/javascript/sorting/cocktail/cocktail_sort.js
export function cocktailShakerSort(values) {
  const a = [...values];
  let start = 0;
  let end = a.length - 1;
  let swapped = true;
  while (swapped) {
    swapped = false;
    for (let i = start; i < end; i++) {
      if (a[i] > a[i + 1]) {
        [a[i], a[i + 1]] = [a[i + 1], a[i]];
        swapped = true;
      }
    }
    if (!swapped) break;
    swapped = false;
    end--;
    for (let i = end; i > start; i--) {
      if (a[i - 1] > a[i]) {
        [a[i - 1], a[i]] = [a[i], a[i - 1]];
        swapped = true;
      }
    }
    start++;
  }
  return a;
}
